File: 161403323李凤明/code/HeapSort.c
/*
利用大顶堆(小顶堆)堆顶记录的是最大关键字(最小关键字)这一特性,
使得每次从无序中选择最大记录(最小记录)变得简单。
首先根据该数组元素构建一个完全二叉树
构造初始堆：
从最后一个非叶节点开始调整，
每次调整都是从父节点、左孩子节点、右孩子节点三者中选择最大者跟父节点进行交换
(交换之后可能造成被交换的孩子节点不满足堆的性质，因此每次交换之后要重新对被交换的孩子节点进行调整)。
有了初始堆之后就可以进行排序了
*/



#include<stdio.h>
typedef int ElementType;
int arr1[11]={0,2,87,39,49,34,62,53,6,44,98};
#define LeftChild(i) (2 * (i) + 1)

void Swap(int* a,int* b){
    int temp=*a;
    *a=*b;
    *b=temp;
}

void PercDown(int  A[], int i, int N){
    int child;
    ElementType Tmp;

    for (Tmp = A[i]; 2*i+1 < N; i = child){
        child = 2*i+1; //注意数组下标是从0开始的，所以左孩子的下标不是2*i
        if (child != N - 1 && A[child + 1] > A[child])
            ++child;                //找到最大的儿子节点
        if (Tmp < A[child])
            A[i] = A[child];
        else
            break;
    }
    A[i] = Tmp;
}

void HeapSort(int A[], int N){
    int i;
    for (i = N / 2; i >= 0; --i)
        PercDown(A, i, N);    //构造堆
    for(i=N-1;i>0;--i)
    {
        Swap(&A[0],&A[i]);        //将最大元素（根）与数组末尾元素交换，从而删除最大元素，重新构造堆
        PercDown(A, 0, i);
    }
}

void Print(int A[],int N){
    int i;
    for(i=0;i<N;i++)
    {
        printf("%d ",A[i]);
    }
}

int main()
{
    int arr[10]={2,87,39,49,34,62,53,6,44,98};
    printf("排序前:");
    Print(arr,10);
    printf("\n");
    HeapSort(arr,10);
    printf("排序后:");
    Print(arr,10);
    printf("\n");
    return 0;
}

File: 161403223庞竹/code/堆/src/HeapSort.java

 
public class HeapSort {
   
  /** 
   * 调整为小顶堆（排序后结果为从大到小）
   * 
   * @param array是待调整的堆数组 
   * @param s是待调整的数组元素的位置
   * @param length是数组的长度
   * 
   */
  public static void heapAdjustS(int[] array, int s, int length) {
    int tmp = array[s];
    int child = 2 * s + 1;// 左孩子结点的位置
    System.out.println("待调整结点为：array[" + s + "] = " + tmp);
    while (child < length) {
      // child + 1 是当前调整结点的右孩子
      // 如果有右孩子且小于左孩子，使用右孩子与结点进行比较，否则使用左孩子
      if (child + 1 < length && array[child] > array[child + 1]) {
        child++;
      }
      System.out.println("将与子孩子 array[" + child + "] = " + array[child] + " 进行比较");
      // 如果较小的子孩子比此结点小
      if (array[s] > array[child]) {
        System.out.println("子孩子比其小，交换位置");
        array[s] = array[child];// 把较小的子孩子向上移动，替换当前待调整结点
        s = child;// 待调整结点移动到较小子孩子原来的位置
        array[child] = tmp;
        child = 2 * s + 1;// 继续判断待调整结点是否需要继续调整 
        if (child >= length) {
          System.out.println("没有子孩子了，调整结束");
        } else {
          System.out.println("继续与新的子孩子进行比较");
        }
    } else {
        System.out.println("子孩子均比其大，调整结束");
        break;// 当前待调整结点小于它的左右孩子，不需调整，直接退出
      }
    }
  }
   
  /** 
   * 调整为大顶堆（排序后结果为从小到大）
   * 
   * @param array是待调整的堆数组 
   * @param s是待调整的数组元素的位置
   * @param length是数组的长度
   * 
   */
  public static void heapAdjustB(int[] array, int s, int length) {
    int tmp = array[s];
    int child = 2 * s + 1;// 左孩子结点的位置
    System.out.println("待调整结点为：array[" + s + "] = " + tmp);
    while (child < length) {
      // child + 1 是当前调整结点的右孩子
      // 如果有右孩子且大于左孩子，使用右孩子与结点进行比较，否则使用左孩子
      if (child + 1 < length && array[child] < array[child + 1]) {
        child++;
      }
      System.out.println("将与子孩子 array[" + child + "] = " + array[child] + " 进行比较");
      // 如果较大的子孩子比此结点大
      if (array[s] < array[child]) {
        System.out.println("子孩子比其大，交换位置");
        array[s] = array[child];// 把较大的子孩子向上移动，替换当前待调整结点
        s = child;// 待调整结点移动到较大子孩子原来的位置
        array[child] = tmp;
        child = 2 * s + 1;// 继续判断待调整结点是否需要继续调整
         
        if (child >= length) {
          System.out.println("没有子孩子了，调整结束");
        } else {
          System.out.println("继续与新的子孩子进行比较");
        }
        // continue;
      } else {
        System.out.println("子孩子均比其小，调整结束");
        break;// 当前待调整结点大于它的左右孩子，不需调整，直接退出
      }
    }
  }
    
  /**
   * 堆排序算法
   * 
   * @param array
   * @param inverse true 为倒序排列，false 为正序排列
   */
  public static void heapSort(int[] array, boolean inverse) {
    // 初始堆
    // 最后一个有孩子的结点位置 i = (length - 1) / 2, 以此向上调整各结点使其符合堆
    System.out.println("初始堆开始");
    for (int i = (array.length - 1) / 2; i >= 0; i--) {
      if (inverse) {
        heapAdjustS(array, i, array.length);
      } else {
        heapAdjustB(array, i, array.length);
      }
    }
    System.out.println("初始堆结束");
    for (int i = array.length - 1; i > 0; i--) {
      // 交换堆顶元素H[0]和堆中最后一个元素
      int tmp = array[i];
      array[i] = array[0];
      array[0] = tmp;
      // 每次交换堆顶元素和堆中最后一个元素之后，都要对堆进行调整
      if (inverse) {
        heapAdjustS(array, 0, i);
      } else {
        heapAdjustB(array, 0, i);
      }
    }
  }
 
  public static void main(String[] args) {
    int[] array = { 49, 38, 65, 97, 76, 13, 27, 49 };
    heapSort(array, false);
    for (int i : array) {
      System.out.print(i + " ");
    }
  }
 
}